python/lsst/ts/ess/common/socket_server.py

```diff
@@ -0,0 +1,164 @@
+        self.log: logging.Logger = logging.getLogger(type(self).__name__)
+        self.command_handler: typing.Optional[AbstractCommandHandler] = None
+
+        super().__init__(
+            name=self.name,
+            host=host,
+            port=port,
+            log=self.log,
+            connect_callback=self.connected_callback,
+            family=family,
+        )
+
+    def set_command_handler(self, command_handler: AbstractCommandHandler) -> None:
+        """Set the command handler instance to use. All code using
+        this SocketServer class must call this at least once before
+        sending commands.
+
+        Parameters
+        ----------
+        command_handler: `AbstractCommandHandler`
+            The command handler instance to use.
+        """
+        self.command_handler = command_handler
+
+    def connected_callback(self, server: SocketServer) -> None:
+        """A client has connected or disconnected."""
+        self.read_loop_task.cancel()
+        if server.connected:
+            self.log.info("Client connected.")
+            self.read_loop_task = asyncio.create_task(self.read_loop())
+        else:
+            self.log.info("Client disconnected.")
+
+    async def write(self, data: dict) -> None:
+        """Write the data appended with a newline character.
+
+        The data are encoded via JSON and then passed on to the StreamWriter
+        associated with the socket.
+
+        Parameters
+        ----------
+        data: `dict`
+            The data to write.
+        """
+        self.log.debug(f"Writing data {data}")
+        st = json.dumps({**data})
+        self.log.debug(st)
+        self.writer.write(st.encode() + tcpip.TERMINATOR)
+        await self.writer.drain()
+        self.log.debug("Done")
+
+    async def read_loop(self) -> None:
+        """Read commands and output replies."""
+        try:
+            self.log.info(f"The read_loop begins connected? {self.connected}")
+            while self.connected:
+                self.log.debug("Waiting for next incoming message.")
+                try:
+                    line = await self.reader.readuntil(tcpip.TERMINATOR)
+                    line = line.decode().strip()
+                    self.log.debug(f"Read command line: {line!r}")
+                    items = json.loads(line)
+                    cmd = items["command"]
+                    kwargs = items["parameters"]
+                    if cmd == "exit":
+                        await self.exit()
+                    elif cmd == "disconnect":
+                        await self.disconnect()
+                    else:
+                        assert self.command_handler is not None
+                        await self.command_handler.handle_command(cmd, **kwargs)
+
+                except asyncio.IncompleteReadError:
+                    self.log.exception("Read error encountered. Retrying.")
+
+        except Exception:
```

Comment:
Please consider catching asyncio.CancelledError and having that log a message and return. I don't think you want a logged exception with traceback for that.

Personally I have found it simpler to allow loops such as these to end "normally" in two different ways: (1) have the while condition go false and (2) allow the loop to be cancelled. If you code carefully enough you can probably ensure it's only ever one or the other. But is it worth the hassle? What if future changes allow the other end condition to occur?

Reply:
Yes. Exactly. With the new except clause you can now quietly and safely terminate the read loop in 2 ways: disconnect the client or cancel the loop.